packages/query-core/src/queryObserver.ts

```diff
@@ -764,3 +787,51 @@ function isStale(
 ): boolean {
   return query.isStaleByTime(options.staleTime)
 }
+
+// this function would decide if we will update the observer's 'current'
+// properties after an optimistic reading via getOptimisticResult
+function shouldAssignObserverCurrentProperties<
+  TQueryFnData = unknown,
+  TError = unknown,
+  TData = TQueryFnData,
+  TQueryData = TQueryFnData,
+  TQueryKey extends QueryKey = QueryKey,
+>(
+  observer: QueryObserver<TQueryFnData, TError, TData, TQueryData, TQueryKey>,
+  optimisticResult: QueryObserverResult<TData, TError>,
+  options: DefaultedQueryObserverOptions<
+    TQueryFnData,
+    TError,
+    TData,
+    TQueryData,
+    TQueryKey
+  >,
+) {
+  // it is important to keep this condition like this for three reasons:
+  // 1. It will get removed in the v5
+  // 2. it reads: don't update the properties if we want to keep the previous
+  // data.
+  // 3. The opposite condition (!options.keepPreviousData) would fallthrough
+  // and will result in a bad decision
```

Comment:
yeah I think you are right, especially when we consider that `createResult` is called during render (via `getOptimisticResult`), so we are actually having an observable side-effect during render.

We're now trying to "counteract" that side-effect by adding more writes during render 😅. Saving `selectResult` and `selectError` was done as a perf optimization, so that select wouldn't needlessly be invoked:

https://github.com/TanStack/query/blob/457723f2cd47a460d351d695d94bff962e7f68e5/packages/query-core/src/queryObserver.ts#L450-L456

For that, it's like priming a cache, so I think it should be fine from react's perspective. But if you have better ideas, I'm all ears. We do have quite good test coverage for that area, because it took some iterations to not get it wrong in various edge cases (and we added a test for each one of them).